### PR TITLE
Add glblub implementation

### PR DESCRIPTION
**Pull request by gitstashpop**
Policy version 32 added support for greatest lower bound (glb) and least
upper bound (lub) calculations for a source and target range. This is a
port to the Go selinux library to provide support for users using a
kernel without glblub support.

Some example glblub calculations:

User Permitted Range | Network Device Label | Computed Label
---------------------|----------------------|----------------
s0-s1:c0.c12         | s0                   | s0
s0-s1:c0.c12         | s0-s1:c0.c1023       | s0-s1:c0.c12
s0-s4:c0.c512        | s1-s1:c0.c1023       | s1-s1:c0.c512
s0-s15:c0,c2         | s4-s6:c0.c128        | s4-s6:c0,c2
s0-s4                | s2-s6                | s2-s4
s0-s4                | s5-s8                | INVALID
s5-s8                | s0-s4                | INVALID

### CODE REVIEW

File: go-selinux/selinux_linux.go [hunk 1]
```diff
@@ -445,6 +457,199 @@ func ComputeCreateContext(source string, target string, class string) (string, e
...
+			len++
+			catRange = true
+		} else if catRange {
+			catRange = false
+			start = true
+			if len > 1 {
+				cats = fmt.Sprintf("%s.c%s", cats, strconv.Itoa(int(r)))
+			} else {
+				cats = fmt.Sprintf("%s,c%s", cats, strconv.Itoa(int(r)))
+			}
```
Comment (gitstashpop) [inline go-selinux/selinux_linux.go:565]: I need to reset len = 0 here

**Comment by mpalmi** [comment]
Just a few nits, and we discussed the string->range/range->string conversion out of band. Feel free to send it up and see what they say. No major issues with the logic, just implementation/approach nits.

File: go-selinux/selinux_linux.go [hunk 2]
```diff
@@ -445,6 +459,198 @@ func ComputeCreateContext(source string, target string, class string) (string, e
...
+
+	/* take the least of the high */
+	outrange.high.sens = uint32(math.Min(float64(s.high.sens), float64(t.high.sens)))
+
+	/* find the intersecting categories */
+	if s.low.cats != nil && t.low.cats != nil {
+		cats := roaring.And(s.low.cats, t.low.cats)
+		if !cats.IsEmpty() {
+			outrange.low.cats = roaring.And(s.low.cats, t.low.cats)
+		}
```
Comment (jbrindle) [inline go-selinux/selinux_linux.go]: Why do you need to call it twice?

File: go-selinux/selinux_linux.go [hunk 3]
```diff
@@ -18,6 +19,7 @@ import (
 	"strings"
 	"sync"
 
+	"github.com/RoaringBitmap/roaring"
```
Comment (gitstashpop) [inline go-selinux/selinux_linux.go]: change to a bitmap that doesn't pull as many deps

File: go-selinux/selinux_linux.go [hunk 4]
```diff
@@ -445,6 +459,198 @@ func ComputeCreateContext(source string, target string, class string) (string, e
...
+	}
+
+	if mlsRange.high == nil {
+		mlsRange.high = mlsRange.low
+	}
+
+	return mlsRange, nil
+}
+
+func bitmapToStr(c *roaring.Bitmap) string {
```
Comment (gitstashpop) [inline go-selinux/selinux_linux.go]: add comment

File: go-selinux/selinux_linux.go [hunk 5]
```diff
@@ -52,13 +54,25 @@ type selinuxState struct {
...
+}
+
 var (
 	// ErrMCSAlreadyExists is returned when trying to allocate a duplicate MCS.
 	ErrMCSAlreadyExists = errors.New("MCS label already exists")
 	// ErrEmptyPath is returned when an empty path has been specified.
 	ErrEmptyPath = errors.New("empty path")
 	// InvalidLabel is returned when an invalid label is specified.
 	InvalidLabel = errors.New("Invalid Label")
+	// ErrNoCommonSens is returned when no common sensitivities exist
```
Comment (mpalmi) [inline go-selinux/selinux_linux.go]: update the comment

**Review by jbrindle — APPROVED** [review]

**Pull request closed without merging**